scanpy/get/aggregated.py

```diff
@@ -0,0 +1,401 @@
+                if weight_value is not None:
+                    weight_value = weight_value[mask]
+            return keys, key_index, df_index, weight_value
+
+        key_value = self._groupby
+        keys, key_index = np.unique(_ndarray_from_seq(key_value), return_inverse=True)
+        df_index = np.arange(len(key_index))
+        if self._weight is None:
+            weight_value = None
+        else:
+            weight_value = self._weight.values[df_index]
+        if self._key_set is not None:
+            keys, key_index, df_index, weight_value = _filter_indices(
+                self._key_set, keys, key_index, df_index, weight_value
+            )
+        self._key_index = key_index  # passed to count and count_mean_var to avoid re-extracting in the latter
+        return keys, key_index, df_index, weight_value
+
+
+def _power(X, power):
+    return X ** power if isinstance(X, np.ndarray) else X.power(power)
+
+
+def _ndarray_from_seq(lst: Sequence):
+    # prevents expansion of iterables as axis
+    n = len(lst)
+    if n > 0 and isinstance(lst[0], cabc.Iterable):
+        arr = np.empty(n, dtype=object)
+        arr[:] = lst
+    else:
+        arr = np.array(lst)
+    return arr
+
+
+def _superset_columns(df: pd.DataFrame, groupby_key: str) -> List[str]:
+    """Find all columns which are a superset of the key column.
+
+    Args:
+        df (pd.DataFrame): DataFrame which contains candidate columns.
+        groupby_key (str): Key for column of which to find superset of columns.
+
+    Returns:
+        List[str]: Superset columns.
+    """
+    columns = []
+    groupy_key_codes = df[groupby_key].astype('category')
+    for key in df:
+        if key != groupby_key:
+            key_codes = df[key].astype('category')
+            if all(
+                [
+                    key_codes[groupy_key_codes == group_key_code].nunique() == 1
+                    for group_key_code in groupy_key_codes
+                ]
+            ):
+                columns += [key]
+    return columns
+
+
+def _df_grouped(df: pd.DataFrame, key: str, key_set: List[str]) -> pd.DataFrame:
+    """Generate a grouped-by dataframe (no aggregation) by a key with columns that are supersets of the key column
+
+    Args:
+        df (pd.DataFrame): DataFrame to be grouped.
+        key (str): Column to be grouped on.
+        key_set (List[str]): values in the `key` column to keep before groupby.
+
+    Returns:
+        pd.DataFrame: Grouped-by Dataframe.
+    """
+    df = df.copy()
+    if key_set is not None:
+        df = df[df[key].isin(key_set)]
+    if df[key].dtype.name == 'category':
+        df[key] = df[key].cat.remove_unused_categories()
+    return df.groupby(key).first()[_superset_columns(df, key)]
+
+
+@singledispatch
+def aggregated(
+    adata: AnnData,
+    by: str,
+    how: Literal['count', 'mean', 'sum', 'count_mean_var'] = 'count_mean_var',
+    groupby_df_key: Literal['obs', 'var'] = 'obs',
```

Comment:
```suggestion
    *,
    groupby_df_key: Literal['obs', 'var'] = 'obs',
```

I'd like to limit how many of these can be positional.